<a href="https://colab.research.google.com/github/kazishuvo22/BigMart_Sales_Prediction/blob/master/BigMart_predict.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **BigMart Sales prediction Dataset**

---

Find out the properties of a product, and store which impacts the sales of a product.

---




In [55]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from google.colab import drive
import glob
%matplotlib inline


In [56]:
train_file="/content/drive/My Drive/cse445 dataset/Train.csv"
test_file="/content/drive/My Drive/cse445 dataset/Test.csv"
submission_file="/content/drive/My Drive/cse445 dataset/Submission.csv"

In [57]:
train_file, test_file, submission_file

('/content/drive/My Drive/cse445 dataset/Train.csv',
 '/content/drive/My Drive/cse445 dataset/Test.csv',
 '/content/drive/My Drive/cse445 dataset/Submission.csv')

In [58]:
train = pd.read_csv(train_file)
test = pd.read_csv(test_file)
submission = pd.read_csv(submission_file)

In [59]:
train

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0              FDA15        9.300  ...  Supermarket Type1          3735.1380
1              DRC01        5.920  ...  Supermarket Type2           443.4228
2              FDN15       17.500  ...  Supermarket Type1          2097.2700
3              FDX07       19.200  ...      Grocery Store           732.3800
4              NCD19        8.930  ...  Supermarket Type1           994.7052
...              ...          ...  ...                ...                ...
8518           FDF22        6.865  ...  Supermarket Type1          2778.3834
8519           FDS36        8.380  ...  Supermarket Type1           549.2850
8520           NCJ29       10.600  ...  Supermarket Type1          1193.1136
8521           FDN46        7.210  ...  Supermarket Type2          1845.5976
8522           DRG01       14.800  ...  Supermarket Type1           765.6700

[8523 rows x 12 columns]

In [60]:
test

Item_Identifier  Item_Weight  ... Outlet_Location_Type        Outlet_Type
0              FDW58       20.750  ...               Tier 1  Supermarket Type1
1              FDW14        8.300  ...               Tier 2  Supermarket Type1
2              NCN55       14.600  ...               Tier 3      Grocery Store
3              FDQ58        7.315  ...               Tier 2  Supermarket Type1
4              FDY38          NaN  ...               Tier 3  Supermarket Type3
...              ...          ...  ...                  ...                ...
5676           FDB58       10.500  ...               Tier 1  Supermarket Type1
5677           FDD47        7.600  ...               Tier 3  Supermarket Type2
5678           NCO17       10.000  ...               Tier 2  Supermarket Type1
5679           FDJ26       15.300  ...               Tier 2  Supermarket Type1
5680           FDU37        9.500  ...               Tier 2  Supermarket Type1

[5681 rows x 11 columns]

In [61]:
submission

Unnamed: 0 Item_Identifier Outlet_Identifier  Item_Outlet_Sales
0              0           FDW58            OUT049        1636.244023
1              1           FDW14            OUT017        1409.355910
2              2           NCN55            OUT010         710.594286
3              3           FDQ58            OUT017        2355.184955
4              4           FDY38            OUT027        5857.916205
...          ...             ...               ...                ...
5676        5676           FDB58            OUT046        2125.984810
5677        5677           FDD47            OUT018        2602.671833
5678        5678           NCO17            OUT045        1832.451358
5679        5679           FDJ26            OUT017        3538.685188
5680        5680           FDU37            OUT045        1281.144462

[5681 rows x 4 columns]

## Combine test and train into one file

To explore data in any data science competetion, it is advisable to append test data to the train data. So, we will combine both train and test to carry out data visualization, feature engineering, one-hot encoding, and label encoding. Later we would split this combined data back to train and test datasets.








In [62]:
train['source']='train'
test['source']='test'
data = pd.concat([train, test],ignore_index=True)
print(train.shape, test.shape, data.shape)

(8523, 13) (5681, 12) (14204, 13)


train dataset has 8523 rows and 12 features and test has 5681 rows and 11 columns. train has 1 extra column which is the target variable. We will predict this target variable for the test dataset later.

In [63]:
data.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
source                        object
dtype: object

In [64]:
data.head()

Item_Identifier  Item_Weight  ... Item_Outlet_Sales  source
0           FDA15         9.30  ...         3735.1380   train
1           DRC01         5.92  ...          443.4228   train
2           FDN15        17.50  ...         2097.2700   train
3           FDX07        19.20  ...          732.3800   train
4           NCD19         8.93  ...          994.7052   train

[5 rows x 13 columns]

##Numerical data summary:

In [65]:
data.describe()

Item_Weight  ...  Item_Outlet_Sales
count  11765.000000  ...        8523.000000
mean      12.792854  ...        2181.288914
std        4.652502  ...        1706.499616
min        4.555000  ...          33.290000
25%        8.710000  ...         834.247400
50%       12.600000  ...        1794.331000
75%       16.750000  ...        3101.296400
max       21.350000  ...       13086.964800

[8 rows x 5 columns]

## Data Cleaning
Missing data can have a severe impact on building predictive models because the missing values might be contain some vital information which could help in making better predictions. So, it becomes imperative to carry out missing data imputation.



### Check missing values:

In [66]:
data.apply(lambda x: sum(x.isnull()))

Item_Identifier                 0
Item_Weight                  2439
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  4016
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales            5681
source                          0
dtype: int64


### Filling missing values


In [67]:
data['Item_Outlet_Sales'].value_counts()

958.7520     17
1342.2528    16
1845.5976    15
703.0848     15
1278.3360    14
             ..
3167.8764     1
2226.4352     1
1684.4740     1
1574.6170     1
6692.6216     1
Name: Item_Outlet_Sales, Length: 3493, dtype: int64

In [68]:
data.Item_Outlet_Sales = data.Item_Outlet_Sales.fillna(data.Item_Outlet_Sales.mean())

In [69]:
data.Item_Outlet_Sales 

0        3735.138000
1         443.422800
2        2097.270000
3         732.380000
4         994.705200
            ...     
14199    2181.288914
14200    2181.288914
14201    2181.288914
14202    2181.288914
14203    2181.288914
Name: Item_Outlet_Sales, Length: 14204, dtype: float64

In [70]:
data['Item_Weight'].value_counts()

17.600    135
12.150    127
10.500    123
13.650    115
11.800    113
         ... 
5.210       7
7.960       7
4.615       6
9.035       6
7.850       6
Name: Item_Weight, Length: 415, dtype: int64

In [71]:
data.Item_Weight = data.Item_Weight.fillna(data.Item_Weight.mean())
data.Item_Weight

0         9.30
1         5.92
2        17.50
3        19.20
4         8.93
         ...  
14199    10.50
14200     7.60
14201    10.00
14202    15.30
14203     9.50
Name: Item_Weight, Length: 14204, dtype: float64

In [72]:

data['Outlet_Size'].value_counts()

Medium    4655
Small     3980
High      1553
Name: Outlet_Size, dtype: int64

In [73]:
data.Outlet_Size = data.Outlet_Size.fillna('Medium')
data.Outlet_Size

0        Medium
1        Medium
2        Medium
3        Medium
4          High
          ...  
14199     Small
14200    Medium
14201    Medium
14202    Medium
14203    Medium
Name: Outlet_Size, Length: 14204, dtype: object

In [74]:
data.apply(lambda x: sum(x.isnull()))

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
source                       0
dtype: int64

In [75]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14204 entries, 0 to 14203
Data columns (total 13 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            14204 non-null  object 
 1   Item_Weight                14204 non-null  float64
 2   Item_Fat_Content           14204 non-null  object 
 3   Item_Visibility            14204 non-null  float64
 4   Item_Type                  14204 non-null  object 
 5   Item_MRP                   14204 non-null  float64
 6   Outlet_Identifier          14204 non-null  object 
 7   Outlet_Establishment_Year  14204 non-null  int64  
 8   Outlet_Size                14204 non-null  object 
 9   Outlet_Location_Type       14204 non-null  object 
 10  Outlet_Type                14204 non-null  object 
 11  Item_Outlet_Sales          14204 non-null  float64
 12  source                     14204 non-null  object 
dtypes: float64(4), int64(1), object(8)
memory usag

### Item type combine:

In [76]:
data['Item_Identifier'].value_counts()


FDM56    10
FDM28    10
NCV06    10
FDK41    10
FDX50    10
         ..
FDL50     7
NCL42     7
DRN11     7
FDM50     7
FDM10     7
Name: Item_Identifier, Length: 1559, dtype: int64

In [77]:
data['Item_Type_Combined'] = data['Item_Identifier'].apply(lambda x: x[0:2])
data['Item_Type_Combined'] = data['Item_Type_Combined'].map({'FD':'Food',
                                                             'NC':'Non-Consumable',
                                                             'DR':'Drinks'})

In [78]:
data['Item_Type_Combined'].value_counts()

Food              10201
Non-Consumable     2686
Drinks             1317
Name: Item_Type_Combined, dtype: int64

## Numerical and One-Hot Coding of Categorical variables

In [79]:
#import library
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
label = LabelEncoder()
#New variable for outlet
data['Outlet'] = label.fit_transform(data['Outlet_Identifier'])
var_mod = ['Item_Fat_Content','Outlet_Location_Type','Outlet_Size','Item_Type_Combined','Outlet_Type','Outlet']
label = LabelEncoder()
for i in var_mod:
    data[i] = label.fit_transform(data[i])

In [80]:
data

Item_Identifier  Item_Weight  ...  Item_Type_Combined  Outlet
0               FDA15         9.30  ...                   1       9
1               DRC01         5.92  ...                   0       3
2               FDN15        17.50  ...                   1       9
3               FDX07        19.20  ...                   1       0
4               NCD19         8.93  ...                   2       1
...               ...          ...  ...                 ...     ...
14199           FDB58        10.50  ...                   1       8
14200           FDD47         7.60  ...                   1       3
14201           NCO17        10.00  ...                   2       7
14202           FDJ26        15.30  ...                   1       2
14203           FDU37         9.50  ...                   1       7

[14204 rows x 15 columns]

In [81]:
#One Hot Coding:
data = pd.get_dummies(data, columns=['Item_Fat_Content','Outlet_Location_Type','Outlet_Size','Outlet_Type','Item_Type_Combined','Outlet'])

In [82]:
data

Item_Identifier  Item_Weight  ...  Outlet_8 Outlet_9
0               FDA15         9.30  ...         0        1
1               DRC01         5.92  ...         0        0
2               FDN15        17.50  ...         0        1
3               FDX07        19.20  ...         0        0
4               NCD19         8.93  ...         0        0
...               ...          ...  ...       ...      ...
14199           FDB58        10.50  ...         1        0
14200           FDD47         7.60  ...         0        0
14201           NCO17        10.00  ...         0        0
14202           FDJ26        15.30  ...         0        0
14203           FDU37         9.50  ...         0        0

[14204 rows x 37 columns]

In [85]:
data.head(6)

Item_Identifier  Item_Weight  Item_Visibility  ... Outlet_7  Outlet_8 Outlet_9
0           FDA15        9.300         0.016047  ...        0         0        1
1           DRC01        5.920         0.019278  ...        0         0        0
2           FDN15       17.500         0.016760  ...        0         0        1
3           FDX07       19.200         0.000000  ...        0         0        0
4           NCD19        8.930         0.000000  ...        0         0        0
5           FDP36       10.395         0.000000  ...        0         0        0

[6 rows x 37 columns]

In [86]:
data.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Item_Outlet_Sales            float64
source                        object
Item_Fat_Content_0             uint8
Item_Fat_Content_1             uint8
Item_Fat_Content_2             uint8
Item_Fat_Content_3             uint8
Item_Fat_Content_4             uint8
Outlet_Location_Type_0         uint8
Outlet_Location_Type_1         uint8
Outlet_Location_Type_2         uint8
Outlet_Size_0                  uint8
Outlet_Size_1                  uint8
Outlet_Size_2                  uint8
Outlet_Type_0                  uint8
Outlet_Type_1                  uint8
Outlet_Type_2                  uint8
Outlet_Type_3                  uint8
Item_Type_Combined_0           uint8
Item_Type_Combined_1           uint8
Item_Type_Combined_2           uint8
O